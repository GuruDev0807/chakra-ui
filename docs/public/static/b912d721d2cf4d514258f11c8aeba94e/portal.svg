<svg width="354" height="240" viewBox="0 0 354 240" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect width="354" height="240" fill="white"/>
<rect opacity="0.9" x="104" y="29" width="10" height="181" fill="#494949"/>
<rect opacity="0.9" x="241" y="29" width="10" height="181" fill="#494949"/>
<rect x="26" y="48" width="78" height="143" fill="#A0AEC0"/>
<rect x="251" y="48" width="78" height="143" fill="#A0AEC0"/>
<rect x="114" y="48" width="27" height="143" fill="#E2E8F0"/>
<rect x="213" y="48" width="28" height="143" fill="#E2E8F0"/>
</svg>
