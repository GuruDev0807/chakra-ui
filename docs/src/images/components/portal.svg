<svg width="400" height="400" viewBox="0 0 400 400" fill="none" xmlns="http://www.w3.org/2000/svg">
<rect width="400" height="400" fill="white"/>
<rect opacity="0.9" x="127" y="109" width="10" height="181" fill="#494949"/>
<rect opacity="0.9" x="264" y="109" width="10" height="181" fill="#494949"/>
<rect x="49" y="128" width="78" height="143" fill="#A0AEC0"/>
<rect x="274" y="128" width="78" height="143" fill="#A0AEC0"/>
<rect x="137" y="128" width="27" height="143" fill="#E2E8F0"/>
<rect x="236" y="128" width="28" height="143" fill="#E2E8F0"/>
</svg>
