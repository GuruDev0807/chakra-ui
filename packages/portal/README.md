# Portal

A wrapper for rendering components in React Portals, with support for nested
portals and stacking. No need to use z-index at all with this portal, that's
right!

## Installation

```sh
yarn add @chakra-ui/portal
```
