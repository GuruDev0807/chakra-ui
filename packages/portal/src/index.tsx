export * from "./Portal.manager"
export * from "./Portal"
