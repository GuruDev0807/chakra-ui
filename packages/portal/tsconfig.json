{
  "extends": "../../tsconfig.json",
  "include": ["src", "stories/portal.stories.tsx"],
  "exclude": ["**/*/*.stories.tsx"]
}
