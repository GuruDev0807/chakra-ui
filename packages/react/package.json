{
  "name": "@chakra-ui/react",
  "version": "1.7.0",
  "description": "Responsive and accessible React UI components built with React and Emotion",
  "main": "dist/chakra-ui-react.cjs.js",
  "module": "dist/chakra-ui-react.esm.js",
  "types": "dist/chakra-ui-react.cjs.d.ts",
  "files": [
    "dist"
  ],
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@chakra-ui/accordion": "1.4.0",
    "@chakra-ui/alert": "1.3.0",
    "@chakra-ui/avatar": "1.3.0",
    "@chakra-ui/breadcrumb": "1.3.0",
    "@chakra-ui/button": "1.5.0",
    "@chakra-ui/checkbox": "1.6.0",
    "@chakra-ui/close-button": "1.2.0",
    "@chakra-ui/control-box": "1.1.0",
    "@chakra-ui/counter": "1.2.0",
    "@chakra-ui/css-reset": "1.1.0",
    "@chakra-ui/editable": "1.3.0",
    "@chakra-ui/form-control": "1.5.0",
    "@chakra-ui/hooks": "1.7.0",
    "@chakra-ui/icon": "1.2.0",
    "@chakra-ui/image": "1.1.0",
    "@chakra-ui/input": "1.3.0",
    "@chakra-ui/layout": "1.5.0",
    "@chakra-ui/live-region": "1.1.0",
    "@chakra-ui/media-query": "1.2.0",
    "@chakra-ui/menu": "1.8.0",
    "@chakra-ui/modal": "1.10.0",
    "@chakra-ui/number-input": "1.3.0",
    "@chakra-ui/pin-input": "1.7.0",
    "@chakra-ui/popover": "1.10.0",
    "@chakra-ui/popper": "2.4.0",
    "@chakra-ui/portal": "1.3.0",
    "@chakra-ui/progress": "1.2.0",
    "@chakra-ui/provider": "1.7.0",
    "@chakra-ui/radio": "1.4.0",
    "@chakra-ui/react-env": "1.1.0",
    "@chakra-ui/select": "1.2.0",
    "@chakra-ui/skeleton": "1.2.0",
    "@chakra-ui/slider": "1.5.0",
    "@chakra-ui/spinner": "1.2.0",
    "@chakra-ui/stat": "1.2.0",
    "@chakra-ui/switch": "1.3.0",
    "@chakra-ui/system": "1.8.0",
    "@chakra-ui/table": "1.3.0",
    "@chakra-ui/tabs": "1.6.0",
    "@chakra-ui/tag": "1.2.0",
    "@chakra-ui/textarea": "1.2.0",
    "@chakra-ui/theme": "1.12.0",
    "@chakra-ui/toast": "1.4.0",
    "@chakra-ui/tooltip": "1.4.0",
    "@chakra-ui/transition": "1.4.0",
    "@chakra-ui/utils": "1.9.0",
    "@chakra-ui/visually-hidden": "1.1.0"
  },
  "peerDependencies": {
    "@emotion/react": "^11.0.0",
    "@emotion/styled": "^11.0.0",
    "framer-motion": "3.x || 4.x || 5.x",
    "react": ">=16.8.6",
    "react-dom": ">=16.8.6"
  },
  "author": "Segun Adebayo <sage@adebayosegun.com>",
  "license": "MIT",
  "homepage": "https://chakra-ui.com/",
  "repository": {
    "type": "git",
    "url": "https://github.com/chakra-ui/chakra-ui",
    "directory": "packages/react"
  },
  "keywords": [
    "react",
    "ui",
    "design-system",
    "react-components",
    "uikit",
    "accessible",
    "components",
    "emotion",
    "library",
    "design-system"
  ],
  "storybook": {
    "title": "Chakra UI",
    "url": "https://chakra-ui.netlify.app"
  },
  "devDependencies": {
    "@emotion/react": "^11.4.0",
    "@emotion/styled": "^11.3.0",
    "framer-motion": "^4.0.0",
    "react": "^17.0.1",
    "react-dom": "^17.0.2"
  }
}
