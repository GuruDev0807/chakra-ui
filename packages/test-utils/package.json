{
  "name": "@chakra-ui/test-utils",
  "version": "1.0.0-next.6",
  "private": true,
  "description": "Testing utilities for Chakra UI's components",
  "keywords": [
    "chakra",
    "test utils",
    "testing",
    "utils",
    "testing library",
    "react testing library"
  ],
  "author": "Segun Adebayo <sage@adebayosegun.com>",
  "homepage": "https://github.com/chakra-ui/chakra-ui#readme",
  "license": "MIT",
  "main": "src/index.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/chakra-ui/chakra-ui.git"
  },
  "bugs": {
    "url": "https://github.com/chakra-ui/chakra-ui/issues"
  },
  "dependencies": {
    "@chakra-ui/system": "^1.0.0-next.6",
    "@chakra-ui/theme": "^1.0.0-next.6",
    "@testing-library/jest-dom": "5.1.1",
    "@testing-library/react": "10.4.5",
    "@testing-library/react-hooks": "3.2.1",
    "@testing-library/user-event": "12.0.11",
    "@types/jest-axe": "3.2.1",
    "jest-axe": "3.4.0",
    "react-test-renderer": "16.12.0"
  },
  "peerDependencies": {
    "react": "16.x"
  }
}
