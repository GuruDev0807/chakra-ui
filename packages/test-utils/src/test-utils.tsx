import { ChakraProvider } from "@chakra-ui/core"
import "@testing-library/jest-dom/extend-expect"
import {
  render as RTLRender,
  RenderOptions,
  fireEvent,
  RenderResult,
  Queries,
} from "@testing-library/react"
import * as React from "react"
import { toHaveNoViolations } from "jest-axe"
import serializer from "jest-emotion"

expect.addSnapshotSerializer(serializer)
expect.extend(toHaveNoViolations)

const AllProviders: React.FC = ({ children }) => (
  <ChakraProvider resetCSS>{children}</ChakraProvider>
)

export function render(
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">,
): RenderResult

export function render(ui: React.ReactElement, options?: RenderOptions) {
  return RTLRender(ui, { wrapper: AllProviders, ...options })
}

export * from "@testing-library/react"

export {
  act as invoke,
  renderHook,
  RenderHookOptions,
  RenderHookResult,
} from "@testing-library/react-hooks"

export { default as userEvent } from "@testing-library/user-event"

export * from "jest-axe"

export const escape = (ui: HTMLElement) =>
  fireEvent.keyDown(ui, { key: "Escape", keyCode: 27 })
