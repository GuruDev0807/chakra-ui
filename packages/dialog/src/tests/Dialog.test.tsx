import * as React from "react"
import { render, userEvent, fireEvent } from "@chakra-ui/test-utils"
import { PortalManager } from "@chakra-ui/portal"
import {
  Dialog,
  DialogBody,
  DialogCloseButton,
  DialogContent,
  DialogFooter,
  DialogHeader,
  DialogOverlay,
} from ".."

const renderWithPortal = (ui: React.ReactElement) =>
  render(<PortalManager>{ui}</PortalManager>)

test("Dialog renders correctly", () => {
  const tools = renderWithPortal(
    <Dialog isOpen onClose={jest.fn()}>
      <DialogOverlay>
        <DialogContent>
          <DialogHeader>Dialog header</DialogHeader>
          <DialogCloseButton />
          <DialogBody>Dialog body</DialogBody>
          <DialogFooter>Dialog footer</DialogFooter>
        </DialogContent>
      </DialogOverlay>
    </Dialog>,
  )
  expect(tools.asFragment()).toMatchSnapshot()
})

test("has the proper aria attributes", () => {
  const tools = renderWithPortal(
    <Dialog isOpen onClose={jest.fn()}>
      <DialogOverlay>
        <DialogContent>
          <DialogHeader>Dialog header</DialogHeader>
          <DialogBody>Dialog body</DialogBody>
        </DialogContent>
      </DialogOverlay>
    </Dialog>,
  )
  // dialog is labelled by the header
  const dialog = tools.getByLabelText("Dialog header")

  expect(dialog).toHaveAttribute("role", "dialog")
  expect(dialog).toHaveAttribute("aria-modal", "true")
  expect(dialog).toHaveAttribute("aria-describedby", "chakra-dialog--body-3")
})

test("clicking the close button calls the onClose callback", () => {
  const onClose = jest.fn()
  const tools = renderWithPortal(
    <Dialog isOpen onClose={onClose}>
      <DialogOverlay>
        <DialogContent>
          <DialogHeader>Dialog header</DialogHeader>
          <DialogCloseButton />
        </DialogContent>
      </DialogOverlay>
    </Dialog>,
  )

  // click the close button
  fireEvent.click(tools.getByLabelText("Close"))

  expect(onClose).toHaveBeenCalled()
})

test('clicking overlay or pressing "esc" calls the onClose callback', () => {
  const onClose = jest.fn()
  const tools = renderWithPortal(
    <Dialog isOpen onClose={onClose}>
      <DialogOverlay data-testid="overlay">
        <DialogContent>
          <DialogHeader>Dialog header</DialogHeader>
          <DialogBody>Dialog body</DialogBody>
        </DialogContent>
      </DialogOverlay>
    </Dialog>,
  )
  const overlay = tools.getByTestId("overlay")

  userEvent.click(overlay)
  fireEvent.keyDown(overlay, { key: "Escape", keyCode: 27 })

  expect(onClose).toHaveBeenCalledTimes(2)
})

test("focuses the initial focus ref when opened", () => {
  const Component = () => {
    const [isOpen, setIsOpen] = React.useState(false)
    const inputRef = React.useRef(null)
    return (
      <>
        <button data-testid="button" onClick={() => setIsOpen(true)}>
          Open
        </button>
        <Dialog isOpen={isOpen} initialFocusRef={inputRef} onClose={jest.fn()}>
          <DialogOverlay>
            <DialogContent>
              <DialogHeader>Dialog header</DialogHeader>
              <DialogBody>
                <input />
                <input />
                <input data-testid="input" ref={inputRef} />
              </DialogBody>
            </DialogContent>
          </DialogOverlay>
        </Dialog>
      </>
    )
  }
  const tools = renderWithPortal(<Component />)

  // click button, opening the modal
  fireEvent.click(tools.getByTestId("button"))

  // input is now the active element
  expect(document.activeElement).toEqual(tools.getByTestId("input"))
})

test("returns focus when closed", () => {
  const Component = () => {
    const [isOpen, setIsOpen] = React.useState(false)
    const buttonRef = React.useRef(null)
    return (
      <>
        <button
          ref={buttonRef}
          data-testid="button"
          onClick={() => setIsOpen(true)}
        >
          Open
        </button>
        <Dialog
          finalFocusRef={buttonRef}
          isOpen={isOpen}
          onClose={() => setIsOpen(false)}
        >
          <DialogOverlay>
            <DialogContent>
              <DialogHeader>Dialog header</DialogHeader>
              <DialogCloseButton />
              <DialogBody>Dialog body</DialogBody>
            </DialogContent>
          </DialogOverlay>
        </Dialog>
      </>
    )
  }
  const tools = renderWithPortal(<Component />)
  const button = tools.getByTestId("button")

  // make sure button isn't focused at the start
  expect(document.activeElement).not.toEqual(button)

  // open and close the modal
  fireEvent.click(button)
  fireEvent.click(tools.getByLabelText("Close"))

  expect(document.activeElement).toEqual(button)
})
