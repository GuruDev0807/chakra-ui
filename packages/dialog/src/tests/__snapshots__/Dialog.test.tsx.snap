// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog renders correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-portal-manager"
  >
    <div
      class="chakra-portal"
    >
      <div
        aria-hidden="true"
        data-aria-hidden="true"
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
      <div
        aria-hidden="true"
        data-aria-hidden="true"
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="1"
      />
      <div
        data-focus-lock-disabled="false"
      >
        <div
          class="css-1klbp1o"
          data-chakra-dialog-overlay=""
        >
          <section
            aria-describedby="chakra-dialog--body-1"
            aria-labelledby="chakra-dialog--header-1"
            aria-modal="true"
            class="css-105rhdj"
            data-chakra-dialog-content=""
            id="chakra-dialog-1"
            role="dialog"
            tabindex="-1"
          >
            <header
              class="css-99ylup"
              data-chakra-dialog-header=""
              id="chakra-dialog--header-1"
            >
              Dialog header
            </header>
            <button
              aria-label="Close"
              class="css-f0tj8y"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="css-9na4ut"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <div
              class="css-qjt6jp"
              data-chakra-dialog-body=""
              id="chakra-dialog--body-1"
            >
              Dialog body
            </div>
            <footer
              class="css-1k0710x"
            >
              Dialog footer
            </footer>
          </section>
        </div>
      </div>
      <div
        aria-hidden="true"
        data-aria-hidden="true"
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
    </div>
  </div>
</DocumentFragment>
`;
